moshi/src/main/java/com/squareup/moshi/JsonAdapter.java

```diff
@@ -37,7 +37,12 @@
   }
 
   @CheckReturnValue public final @Nullable T fromJson(String string) throws IOException {
-    return fromJson(new Buffer().writeUtf8(string));
+    JsonReader reader = JsonReader.of(new Buffer().writeUtf8(string));
+    T result = fromJson(reader);
+    if (reader.peek() != JsonReader.Token.END_DOCUMENT) {
```

Comment:
good call on making lenient adapters skip this check. we would need the lenient state, anyway, due to to the reading here.